### PR TITLE
Enhance Arizona parse transform with performance and robustness improvements

### PR DESCRIPTION
**Pull request by williamthome**
# Description

Depends on #329.

## Summary

This PR significantly improves the `arizona_parse_transform` module with comprehensive performance optimizations, robustness enhancements, and documentation accuracy fixes. All improvements maintain full backward compatibility while delivering substantial performance gains and safer error handling.

### **Faster Function Validation**

- **Before**: O(n×m) complexity using `lists:member/2` for each function lookup
- **After**: O(1) complexity using `sets:is_element/2` with pre-built sets
- **Impact**: Dramatically faster compilation for modules with many functions

### **Memory Reduction in AST Processing**

- **Before**: 3 separate full AST traversals for extracting exports, definitions, and arizona functions
- **After**: Single-pass extraction with `extract_functions_single_pass/1`
- **Impact**: Significant memory savings during compilation of large modules

### **Safe Binary Operations**

- Added `safe_is_uppercase_variable/1` with bounds checking for empty binaries
- Prevents crashes when processing malformed or edge-case variable names
- **Before**: `binary:first(VarName)` could crash on empty binaries
- **After**: Graceful handling with proper validation

### **Exception Handling for Unsafe Operations**

- Wrapped `erl_eval:expr/2` in `safe_eval_binary_expr/1` with try/catch blocks
- **Before**: Template extraction failures could crash compilation
- **After**: Graceful error handling with detailed error messages

### **Consistent Error Handling**

- All errors now use consistent `error(Reason, none, error_info(Info))` pattern
- Added `template_extraction_failed` error type to `format_error/2`
- Enhanced error messages provide actionable guidance for developers

### **Corrected README Examples**
- **Before**: Misleading examples showing "variable extraction" that didn't match implementation
- **After**: Accurate examples demonstrating actual dependency tracking behavior
- Fixed terminology throughout: "variable extraction" → "dependency tracking"

### **Updated Module Documentation**
- Corrected moduledoc to accurately reflect compile-time optimization behavior
- Removed misleading claims about variable assignment support
- Added clear explanations of selective processing requirements

### **Comprehensive Edge Case Coverage**
- Added 4 new test cases (347 total tests, all passing)
- `test_empty_binary_variable_validation`: Tests safety improvements
- `test_safe_template_extraction_error_handling`: Tests error handling robustness  
- `test_single_pass_function_extraction_performance`: Tests performance optimizations
- `test_template_extraction_failed_error_formatting`: Tests error formatting

### **Coverage Maintained**
- **arizona_parse_transform**: 93% coverage
- **Total project**: 95% coverage
- All CI checks passing: formatting, linting, xref, dialyzer, tests

### **Modern Erlang Practices**
- Use `try/catch` instead of deprecated `catch` syntax
- Use `~""` binary syntax instead of `<<>>` where appropriate
- Refactored long functions to meet Elvis linting standards (<30 lines per clause)

### **Maintainable Architecture**
- Split `validate_arizona_functions/3` into smaller, focused functions
- Clear separation of concerns between validation, extraction, and transformation
- Consistent error handling patterns throughout the module

## 🔄 Breaking Changes

**None** - This PR maintains full backward compatibility. All existing code using arizona_parse_transform will continue to work without changes.

## 🎯 Impact

This PR transforms the arizona_parse_transform from a functional but suboptimal implementation into a production-ready, high-performance module suitable for large-scale applications. The improvements directly address the technical debt identified in previous analysis while maintaining the clean architecture and excellent error reporting that makes Arizona a joy to use.

---

- [x] I have performed a self-review of my changes
- [x] I have read and understood the [contributing guidelines](/arizona-framework/arizona/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: rebar.config [hunk 1]
```diff
@@ -46,7 +46,7 @@
         ex_doc
     ]},
     {ci, [
-        {do, "default as test check"},
+        check,
```
Comment (williamthome) [inline rebar.config:49]: CI wasn't failing when using the test profile, but by using the default, it detected the `xref` errors in `arizona_parse_transform`.

Comment (paulo-ferraz-oliveira) [reply rebar.config:49]: Why not avoid the module, simply? I'm not really sure of the context here, though.

**Review by paulo-ferraz-oliveira — APPROVED** [review]

File: rebar.config (hunk 1 above)
Comment (williamthome) [reply rebar.config:49]: See https://github.com/arizona-framework/arizona/pull/334/files#r2206140796

File: src/arizona_parse_transform.erl [hunk 2]
```diff
@@ -73,13 +34,20 @@ them with optimized versions that avoid runtime template parsing overhead.
...
 %% --------------------------------------------------------------------
 
 -ignore_xref([parse_transform/2]).
 -ignore_xref([format_error/2]).
+-ignore_xref([extract_arizona_functions/1]).
+-ignore_xref([analyze_function_for_bindings/1]).
+-ignore_xref([generate_vars_indexes/2]).
+-ignore_xref([parse_template_for_stateful_with_context/5]).
+-ignore_xref([build_function_bindings_map/2]).
+-ignore_xref([extract_functions_single_pass/1]).
```
[the remark this answers is not in the document]
Comment (paulo-ferraz-oliveira) [reply src/arizona_parse_transform.erl:50]: 👍
